### PR TITLE
Remove weak-refs to the compilation (without generated docs) in the compilation tracker.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
A/B (2500 people in each cohort) showed no statistical significance to this either and it adds a lot of complexity.  

Theory here is that generally the compilations we end up making we end up needing.  So better to hold onto them.  Compilations that are not needed (like from prior snapshots) naturally just get GC'ed.  So this was only about being able to let go more aggressively compilations from latest snapshot.  However, as things are continually changing, and those are being let go anyways, it's not really important to do this.

Should be read commit by commit.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.CompilationTrackerState.cs
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.CompilationTrackerState.cs:181]: effectively DisableCompilationTrackerWeakCompilationReferences becomes 'true'.  so this entire method just became `return new ConstantValueSource...`.  so that got inlined at all callsites.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.CompilationTrackerState.cs
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.CompilationTrackerState.cs:115]: after the first commit, everything that passed along a value to this was a ConstantValueSource.  So we can just remove the indirection.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs:79]: note: i'm going to make an A/B where we do this for FinalCompilationWithGeneratedDocuments as well.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs:685]: Can this calling function also remove the solutionServices parameter?

**Review by jasonmalinowski — APPROVED** [review]
And There Was Much Rejoicing

The only thing I'd request changed in this PR was the FinalState has duplicate parameters, the "source" can be removed since the compilation is also being passed directly at that point. Should be an easy fix.

More generally, I do wonder what else we can also delete at this point; I'm going to make a note for myself to stare at what's left after this is merged and figure out if we can clean up a lot more still. But that's a much more expensive analysis that isn't worth blocking this on.

**Comment by jasonmalinowski** [comment]
(and BTW thanks for the commits being broken apart, really helped in this PR)

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.CompilationTrackerState.cs
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.CompilationTrackerState.cs]: > The only thing I'd request changed in this PR was the FinalState has duplicate parameters, the "source" can be removed since the compilation is also being passed directly at that point. Should be an easy fix.

Done.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs:79]: Hrmm... i don't see how.  But i don't know enough about this code to now for certain.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: done!

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs:272]: i have.  and the FinalCompilationWithGeneratedDocuments is non-nullable.  the contract call is to really just have us die if we're wrong.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs:685]: nope!  still used elsewhere in this method.

**Pull request merged**
